### PR TITLE
fix(gatsby-remark-images): GATSBY_EMPTY_ALT figcaption generation

### LINKED ISSUE #30285: Using GATSBY_EMPTY_ALT ignored for figcaption-generation by gatsby-remark-image ie creates non-sense figcaption
The fix described in issue #20179 introduces a regression, especially from an accessibility perspective. When declaring an empty alt-attribute for an image using the keyword `GATSBY_EMPTY_ALT` in markdown (`GATSBY_EMPTY_ALT` *and* having set `showCaptions: true` in the plugin-options in `gatsby-config.js`, the plugin produces unwanted HTML:

```html
<figure class="gatsby-resp-image-figure" ...>
    <span class="gatsby-resp-image-wrapper" style="...">
      <span class="gatsby-resp-image-background-image" style="..."></span>
        <img class="gatsby-resp-image-image" alt="" title="" src="/path.jpg" ...>
    </span>
</figure>
<figcaption class="gatsby-resp-image-figcaption"><p>GATSBY<em>EMPTY</em>ALT</p></figcaption>
```

* the package produces the empty value for the `alt`-attribute value of the `img`-element like instructed (line 4);
* but it inserts an (obviously) unwanted  `figcaption`-element below the `figure`-element (line 7) that contains unwanted (and non-sense) content:  `<p>GATSBY<em>EMPTY</em>ALT</p>`.
 
This seems to fix the problem for current 4.x versions:

```diff
--- unpatched.index.js	2021-03-15 10:09:19.898810753 +0100
+++ patched.index.js	2021-03-15 10:11:28.041596794 +0100
@@ -114,7 +114,7 @@
           switch (_context.prev = _context.next) {
             case 0:
               getCaptionString = function getCaptionString() {
-                var captionOptions = Array.isArray(options.showCaptions) ? options.showCaptions : options.showCaptions === true ? ["title", "alt"] : false;
+                var captionOptions = Array.isArray(options.showCaptions) ? options.showCaptions : ( options.showCaptions === true && node.alt !== EMPTY_ALT) ? ["title", "alt"] : false;
 
                 if (captionOptions) {
                   for (var _iterator = _createForOfIteratorHelperLoose(captionOptions), _step; !(_step = _iterator()).done;) {
```

Furthermore, this feature should be documented in `README.md:

```diff
--- /tmp/README.md	2021-03-15 16:05:28.328713216 +0100
+++ node_modules/gatsby-remark-images/README.md	2021-03-15 16:12:26.848647108 +0100
@@ -50,8 +50,7 @@
By default, the text `Alt text here` will be used as the alt attribute of the generated `img` tag
-.
+ and for the contents of the generated `figcaption` element.
If an empty alt attribute like `alt=""`
- is wished,
+ and no figcaption is wished,
a reserved keyword `GATSBY_EMPTY_ALT` can be used.
```
 
```diff
--- /tmp/README.md	2021-03-15 16:05:28.328713216 +0100
+++ node_modules/gatsby-remark-images/README.md	2021-03-15 16:12:26.848647108 +0100
@@ -63,7 +62,7 @@
| `showCaptions`          | `false` | Add a caption to each image with the contents of the title attribute, when this is not empty. 
...
If you just want to use the title (and omit captions for images that have alt attributes but no title), pass `['title']`. 
+ If the alt tag contains the reserved `GATSBY_EMPTY_ALT` value no `figcaption` will be produced. |
```

_Originally posted by @ronalde in https://github.com/gatsbyjs/gatsby/issues/20179#issuecomment-798720484_

### PR DESCRIPTION
**Pull request by nategiraudeau**
## Description

#30285

### Documentation

Updated README.

## Related Issues

Fixes #30285.

### CODE REVIEW

**Review by ronalde — APPROVED** [review]
Great!

File: packages/gatsby-remark-images/src/index.js [hunk 1]
```diff
@@ -83,9 +83,9 @@ module.exports = (
     const getCaptionString = () => {
       const captionOptions = Array.isArray(options.showCaptions)
         ? options.showCaptions
-        : options.showCaptions === true
-        ? [`title`, `alt`]
-        : false
+        : (options.showCaptions === true && node.alt !== EMPTY_ALT)
+          ? ["title", "alt"]
+          : false
```
Comment (vladar) [inline packages/gatsby-remark-images/src/index.js:88]: This should satisfy linter:
```suggestion
        : options.showCaptions === true && node.alt !== EMPTY_ALT
        ? [`title`, `alt`]
        : false
```

Comment (nategiraudeau) [reply packages/gatsby-remark-images/src/index.js]: Thanks.

Comment (ronalde) [reply packages/gatsby-remark-images/src/index.js]: @vladar and @nategiraudeau: the linter is still not satisfied, could you please have another look?

Comment (nategiraudeau) [reply packages/gatsby-remark-images/src/index.js]: I'm just waiting for it to get approved.

File: packages/gatsby-remark-images/src/index.js [hunk 2]
```diff
@@ -83,7 +83,7 @@ module.exports = (
     const getCaptionString = () => {
       const captionOptions = Array.isArray(options.showCaptions)
         ? options.showCaptions
-        : options.showCaptions === true
+        : options.showCaptions === true && node.alt !== EMPTY_ALT
```
Comment (motou) [inline packages/gatsby-remark-images/src/index.js:86]: This line introduces an unwanted behavior for the generation of image title. Actually, the logic to detect `EMPTY_ALT` should be put after line 98.

**Comment by nategiraudeau** [comment]
@vladar I finally fixed the linter problem.

**Comment by ronalde** [comment]
Still no luck @nategiraudeau?

**Comment by nategiraudeau** [comment]
The tests seem to all be passing. Just waiting for someone to review.

**Review by wardpeet — APPROVED** [review]

**Pull request merged**